skcosmo/feature_selection/voronoi_fps.py

```diff
@@ -0,0 +1,229 @@
+            self.norms_[self.selected_idx_[: self.n_selected_]]
+            + self.norms_[last_selected]
+            - 2 * (self.X_selected_[:, : self.n_selected_].T @ X[:, last_selected])
+        ) * 0.25
+
+        """
+        These are the points for which need to recompute distances.
+        Let:
+        L is the last point selected;
+        S are the selected points from before this iteration;
+        X are the candidates;
+        The logic here is that we want to check if d(XL) can be smaller than
+        min(d(XS)) (which is stored in self.haussdorf_)
+        now, if a point belongs to the Voronoi cell of S then min(d(XS_i))=d(XS).
+        Triangle inequality implies that d(XL)>=|d(XS) - d(SL)| so we just need to
+        check if |d(XS) - d(SL)|>= d(XS) to know that we don't need to check X.
+        but |d(XS) - d(SL)|^2>= d(XS)^2 if and only if d(SL)/2 > d(SX)
+        """
+        active_points = np.where(
+            self.sel_d2q_[self.vlocation_of_idx] < self.haussdorf_
+        )[0]
+
+        return active_points
+
+    def _calculate_distances(self, X, last_selected, **kwargs):
+
+        # n_selected has not been incremented, so index of new voronoi is
+        # n_selected
+
+        if self.n_selected_ == 0:
+            self.haussdorf_ = super()._calculate_distances(X, last_selected)
+            # tracker of how many distances must be computed at each step
+            updated_points = np.arange(X.shape[-1], dtype=int)
+        else:
+            active_points = self._get_active(X, last_selected)
+
+            # we need to compute distances between the new point and all the points
+            # in the active Voronoi cells.
+            if len(active_points) > 0:
+                if len(active_points) / X.shape[1] > self.full_fraction:
+                    # if the number of distances we need to compute is large, it is
+                    # better to switch to a full-matrix-algebra calculation.
+                    self.new_dist_[:] = super()._calculate_distances(X, last_selected)
+                else:
+                    # ... else we only iterate over the active points, although
+                    # this involves more memory jumps, and can be more costly than
+                    # computing all distances.
+                    self.new_dist_[:] = self.haussdorf_
+
+                    self.new_dist_[active_points] = (
+                        self.norms_[active_points]
+                        + self.norms_[last_selected]
+                        - 2 * X[:, last_selected] @ X[:, active_points]
+                    )
+                    self.new_dist_[last_selected] = 0
+                # updates haussdorf distances and keeps track of the updated points
+                updated_points = np.where(self.new_dist_ < self.haussdorf_)[0]
+                np.minimum(self.haussdorf_, self.new_dist_, self.haussdorf_)
+            else:
+                updated_points = np.array([])
```

Comment:
Could we add tests covering edge cases, such as an empty list of point and selecting all points in the system (which this line of code seems to be dealing with)?